include/fmt/core.h

```diff
@@ -1316,6 +1382,38 @@ struct checked_args : format_arg_store<
   basic_format_args<context> operator*() const { return *this; }
 };
 
+/**
+  \rst
+  The function ``to_string_view`` adapts non-intrusively any kind of string or
+  string-like type if the user provides a (possibly templated) overload of
+  ``to_string_view`` which takes an instance of the string class
+  ``StringType<char_type>`` and returns a ``fmt::basic_string_view<char_type>``.
+  The conversion function must live in the very same namespace as
+  ``StringType<char_type>`` to be picked up by ADL. Non-templated string types
+  like f.e. QString must return a ``basic_string_view`` with a fixed matching
+  char type.
+
+  **Example**::
+
+    namespace ofStringType {
+
+    template <typename Char>
+    inline basic_string_view<char_type> to_string_view(const StringType<Char> &S) {
+        return { S.data(), S.length() };
+    }
+
+    or
+
+    inline basic_string_view<wchar_t> to_string_view(const WideStringType &S) {
+        return { S.data(), S.length() };
+    }
+
+    }
+
+    std::string message = fmt::format(StringType<char>("The answer is {}"), 42);
+  \endrst
+ */
+
 template <typename S>
 inline basic_string_view<FMT_CHAR(S)> to_string_view(const S &s) {
```

Comment:
Now that `to_string_view` is an extension point, it should live in the `fmt` namespace. This will also eliminate the need for `using internal::to_string_view` and, likely, the `internal2` namespace.

Reply:
Hi Victor!
Actually, the fact that the built-in `to_string_view` sits in a **sibling** namespace to `internal2` is the gist of the magic that makes the whole kaboodle work in the first place! In fact, the ADL detection would work much easier without all of this crazy stuff if the detection class could sit in a sibling namespace of `fmt` itself (which it can't). As soon as you move `internal::to_string_view` and the `internal2` detection logic into the same namespace (or the detection logic into a child namespace such as `internal`), the ADL search would pick up the built-in `to_string_view` as possibly viable candidate again if a compiler finds some acceptable conversion path. The whole point of the ADL logic is to find only as **exact** matches as possible without competing overloads.

Did you happen to run a quick test on your machine with your suggested changes? In particular with some of the mentioned problematic compilers?

Reply:
> Did you happen to run a quick test on your machine with your suggested changes? In particular with some of the mentioned problematic compilers?

Hmm, it does break the legacy compilers. There is also some redundancy between `to_string_view` and `format_string_traits`. Let me take a look in more details and report back.